### ISSUE TITLE
float32 matrix dot product has different precision in CPU/GPU

### ISSUE THREAD

**Issue by cow8** [pytorch/pytorch#79270]
### 🐛 Describe the bug

sample code:
```
import numpy as np
from scipy.spatial.transform import Rotation
import torch
import scipy
print(torch.__version__)
print(scipy.__version__)
print(np.__version__)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

def test(dtype,device,pts):
    R = torch.tensor(
        Rotation.from_euler("xy", [0, 0], degrees=True).as_matrix().astype(np.float64),
        dtype=dtype,
        device=device,
    )
    t = torch.tensor(np.array([0, 0, 1e3]), dtype=dtype, device=device)
    
    pts1 = torch.tensor(pts, dtype=dtype, device=device)
    pts2 = (R.T @ ((pts1 - (t[None, :])).T)).T
    pts3 = (R @ (pts2.T)).T + t[None, :]
    print(((pts1 -pts3)**2).sum(-1).mean())

pts = 500*np.random.randn(1024,3).astype(np.float32)
test(torch.float16,"cuda",pts)
test(torch.float16,"cpu",pts)
test(torch.float32,"cuda",pts)
test(torch.float32,"cpu",pts)
test(torch.float64,"cuda",pts)
test(torch.float64,"cpu",pts)
```

output:
```
1.10.0
1.8.1
1.22.3
tensor(0.0645, device='cuda:0', dtype=torch.float16)
tensor(0.0645, dtype=torch.float16)
tensor(0.0737, device='cuda:0')
tensor(1.0279e-09)
tensor(0., device='cuda:0', dtype=torch.float64)
tensor(0., dtype=torch.float64)
```

is it acceptable?

### Versions

print(torch.__version__)
print(scipy.__version__)
print(np.__version__)
1.10.0
1.8.1
1.22.3

cc @zasdfgbnm @ptrblck

**Comment by xsacha** [comment]
Is TF32 enabled for CUDA? Default in some versions I think.

**Comment by cow8** [comment]
@xsacha IT IS! thanks  a lot!
update dated sample code:
```
import numpy as np
from scipy.spatial.transform import Rotation
import torch
import scipy
print(torch.__version__)
print(scipy.__version__)
print(np.__version__)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

def test(dtype,device,pts):
    R = torch.tensor(
        Rotation.from_euler("xy", [0, 0], degrees=True).as_matrix().astype(np.float64),
        dtype=dtype,
        device=device,
    )
    t = torch.tensor(np.array([0, 0, 1e3]), dtype=dtype, device=device)
    
    pts1 = torch.tensor(pts, dtype=dtype, device=device)
    pts2 = (R.T @ ((pts1 - (t[None, :])).T)).T
    pts3 = (R @ (pts2.T)).T + t[None, :]
    print(((pts1 -pts3)**2).sum(-1).mean())


print("*"*10,"when enable TF32","*"*10)

# The flag below controls whether to allow TF32 on matmul. This flag defaults to True.
torch.backends.cuda.matmul.allow_tf32 = True

# The flag below controls whether to allow TF32 on cuDNN. This flag defaults to True.
torch.backends.cudnn.allow_tf32 = True
pts = 500*np.random.randn(1024,3).astype(np.float32)
test(torch.float16,"cuda",pts)
test(torch.float16,"cpu",pts)
test(torch.float32,"cuda",pts)
test(torch.float32,"cpu",pts)
test(torch.float64,"cuda",pts)
test(torch.float64,"cpu",pts)

print("*"*10,"when disable TF32","*"*10)

# The flag below controls whether to allow TF32 on matmul. This flag defaults to True.
torch.backends.cuda.matmul.allow_tf32 = False

# The flag below controls whether to allow TF32 on cuDNN. This flag defaults to True.
torch.backends.cudnn.allow_tf32 = False
torch.backends.cudnn.allow_tf32 = True
pts = 500*np.random.randn(1024,3).astype(np.float32)
test(torch.float16,"cuda",pts)
test(torch.float16,"cpu",pts)
test(torch.float32,"cuda",pts)
test(torch.float32,"cpu",pts)
test(torch.float64,"cuda",pts)
test(torch.float64,"cpu",pts)
```

updated sample output:
```
1.10.0
1.8.1
1.22.3
********** when enable TF32 **********
tensor(0.0597, device='cuda:0', dtype=torch.float16)
tensor(0.0597, dtype=torch.float16)
tensor(0.0712, device='cuda:0')
tensor(8.4135e-10)
tensor(0., device='cuda:0', dtype=torch.float64)
tensor(0., dtype=torch.float64)
********** when disable TF32 **********
tensor(0.0709, device='cuda:0', dtype=torch.float16)
tensor(0.0709, dtype=torch.float16)
tensor(1.1154e-09, device='cuda:0')
tensor(1.1154e-09)
tensor(0., device='cuda:0', dtype=torch.float64)
tensor(0., dtype=torch.float64)
```

**Comment by xsacha** [comment]
Looks good :)
The A10 is an awesome GPU for training too :)

The accuracy issue is why TF32 is now disabled by default :(
